File: 300/331 - Verify Preorder Serialization of a Binary Tree.java
import java.util.ArrayList;

public class Solution {
    public boolean isValidSerialization(String preorder) {
        String[] words = preorder.split(",");
        ArrayList<Boolean> list = new ArrayList<Boolean>();
        int len;
        
        for (String s : words) {
            list.add(s.equals("#"));
            while ((len = list.size()) >= 3 && list.get(len - 1) && list.get(len - 2) && ! list.get(len - 3)) {
                list.remove(len - 1);
                list.remove(len - 2);
                list.remove(len - 3);
                list.add(true);
            }
        }
        
        return list.size() == 1 && list.get(0);
    }
}

File: csharp/331.verify-preorder-serialization-of-a-binary-tree.cs
/*
 * @lc app=leetcode id=331 lang=csharp
 *
 * [331] Verify Preorder Serialization of a Binary Tree
 */
public class Solution {

    private int GetRange(string[] split, int i) {
        if (i >= split.Length) {
            return -1;
        }
        if (split[i] == "#") {
            return i + 1;
        }
        i = GetRange(split, i + 1);
        if (i == -1) {
            return -1;
        }
        return GetRange(split, i);
    }

    public bool IsValidSerialization(string preorder) {
        string[] split = preorder.Split(',');
        return GetRange(split, 0) == split.Length;
    }
}

